File: spi.py
from enum import Enum
import re
import sys


class TokenTypes(Enum):
    # Math Operations
    REAL_CONST = r'\d+\.\d+'
    INTEGER_CONST = r'\d+'
    PLUS = r'\+'
    MINUS = r'\-'
    MULTIPLY = r'\*'
    LPAREN = r'\('
    RPAREN = r'\)'
    INTEGER_DIV = r'(?i)DIV'
    FLOAT_DIV = r'/'

    DOT = r'\.'
    ASSIGN = r':='
    ID = r'_?[A-z][A-z0-9]*'
    SEMI = r';'
    COLON = r':'
    COMMA = r','

    # Reserved keywords
    BEGIN = r'(?i)BEGIN'
    END = r'(?i)END'
    PROGRAM = r'(?i)PROGRAM'
    VAR = r'(?i)VAR'

    # types
    INTEGER = r'(?i)INTEGER'
    REAL = r'(?i)REAL'

    COMMENT = r'{[^}]*}'
    EOF = r'$^'

    def __init__(self, regex):
        self.regex = re.compile(regex)

    def __repr__(self):
        return '<TokenType.{}>'.format(self.name)

RESERVED_KEYWORDS = ['BEGIN', 'END', 'PROGRAM', 'DIV', 'VAR', 'INTEGER',
                     'REAL']


class Token:
    def __init__(self, token_type, value):
        assert token_type in TokenTypes or token_type is None
        self.token_type = token_type
        self.value = value

    def __str__(self):
        return 'Token({0.token_type}, {0.value!r})'.format(self)

    def __repr__(self):
        return self.__str__()


class Lexer:
    def __init__(self, text):
        self.text = self._current_text = text

    def skip_comments_whitespace(self):
        text = self._current_text.lstrip()
        match = re.match(r'{[^}]*}', text)
        if match:
            self._current_text = text[match.end():]
            self.skip_comments_whitespace()
        else:
            self._current_text = text

    def get_next_token(self):
        self.skip_comments_whitespace()

        text = self._current_text
        for token_type in TokenTypes:
            match = token_type.regex.match(text)
            if match:
                matched_text = match.group()
                if token_type == TokenTypes.ID:
                    matched_text = matched_text.upper()

                if token_type == TokenTypes.ID and \
                   matched_text in RESERVED_KEYWORDS:
                    continue

                self._current_text = text[match.end():]
                return Token(token_type, matched_text)
        if len(text) > 0:
            raise Exception("Couldn't tokenize {}".format(text))
        return Token(TokenTypes.EOF, None)


class ASTNode:
    def __init__(self, token=None):
        self.token = token


class ProgramNode(ASTNode):
    def __init__(self, name, block):
        self.name = name
        self.block = block


class BlockNode(ASTNode):
    def __init__(self, declarations, compound_statement):
        self.declarations = declarations
        self.child = compound_statement


class VariableDeclarationNode(ASTNode):
    def __init__(self, var_node, type_node):
        self.var_node = var_node
        self.type_node = type_node


class TypeNode(ASTNode):
    pass


class BinaryOpNode(ASTNode):
    def __init__(self, left, op, right):
        super().__init__(op)
        self.left = left
        self.right = right

    def __repr__(self):
        return "<BinaryOp({0.left},{0.token.value},{0.right})>".format(self)


class UnaryOpNode(ASTNode):
    def __init__(self, op, child):
        super().__init__(op)
        self.child = child

    def __repr__(self):
        return "<UnaryOp({0.token.value},{0.child})>".format(self)


class NumNode(ASTNode):
    def __init__(self, token):
        super().__init__(token)

    def __repr__(self):
        return "<Var({0.token.value})>".format(self)

    @property
    def value(self):
        return self.token.value


class CompoundNode(ASTNode):
    def __init__(self, children):
        super().__init__(None)
        self.children = children

    def __repr__(self):
        return "<Compound({0.children!s})>".format(self)


class AssignNode(ASTNode):
    def __init__(self, lvalue, op, rvalue):
        super().__init__(op)
        self.lvalue = lvalue
        self.rvalue = rvalue

    def __repr__(self):
        return "<Assign({0.lvalue},{0.token.value},{0.rvalue})>".format(self)


class VarNode(ASTNode):
    def __init__(self, token):
        super().__init__(token)
        self.token = token

    def __repr__(self):
        return "<Var({0.token.value})>".format(self)


class NoOpNode(ASTNode):
    def __init__(self):
        super().__init__(None)

    def __repr__(self):
        return '<NoOp>'


class Parser:
    '''
    program : compound_statement DOT

    compound_statement : BEGIN statement_list END

    statement_list : statement | statement SEMI statement_list

    statement : compound_statement | assignment_statement | empty

    assignment_statement : variable ASSIGN expr

    empty :

    expr : term ((PLUS|MINUS) term)*

    term : factor ((MULT|DIVIDE) factor)*

    factor : PLUS factor | MINUS factor | INTERGER
            | LPAREN expr RPAREN | variable

    variable : ID
    '''
    def __init__(self, lexer):
        self._lexer = lexer

    def _eat(self, previous=None):
        output = self._current_token
        if previous is not None and \
           output.token_type != previous and \
           (not isinstance(previous, list) or output.token_type not in previous):
            raise Exception("Expected to see {0!r}, found {1!r} instead".
                            format(previous, output))
        self._current_token = self._lexer.get_next_token()
        return output

    def _program(self):
        self._eat(TokenTypes.PROGRAM)
        name = self._variable()
        self._eat(TokenTypes.SEMI)
        block = self._block()
        self._eat(TokenTypes.DOT)
        return ProgramNode(name, block)

    def _block(self):
        """block : declarations compound_statement"""
        declarations = self._declarations()
        compound = self._compound_statement()
        return BlockNode(declarations, compound)

    def _declarations(self):
        """declarations : VAR (variable_declaration SEMI)+ | empty"""
        declarations = []
        if self._current_token.token_type == TokenTypes.VAR:
            self._eat()
            while self._current_token.token_type == TokenTypes.ID:
                var_decl = self._variable_declaration()
                declarations.extend(var_decl)
                self._eat(TokenTypes.SEMI)
            if len(declarations) == 0:
                raise Exception("VAR detected, but no variable declarations")
        return declarations

    def _variable_declaration(self):
        """variable_declaration : ID (COMMA ID)* COLON type_spec"""
        var_nodes = [VarNode(self._eat(TokenTypes.ID))]

        while self._current_token.token_type == TokenTypes.COMMA:
            self._eat()
            var_nodes.append(VarNode(self._eat(TokenTypes.ID)))

        self._eat(TokenTypes.COLON)

        type_node = self._type_spec()

        return list(VariableDeclarationNode(var, type_node)
                    for var in var_nodes)

    def _type_spec(self):
        """type_spec : INTEGER | REAL"""
        return TypeNode(self._eat([TokenTypes.INTEGER, TokenTypes.REAL]))

    def _compound_statement(self):
        self._eat(TokenTypes.BEGIN)
        node = self._statement_list()
        self._eat(TokenTypes.END)
        return node

    def _statement_list(self):
        statements = [self._statement()]
        while self._current_token.token_type == TokenTypes.SEMI:
            self._eat(TokenTypes.SEMI)
            statements.append(self._statement())
        return CompoundNode(statements)

    def _statement(self):
        if self._current_token.token_type == TokenTypes.BEGIN:
            return self._compound_statement()
        elif self._current_token.token_type == TokenTypes.ID:
            return self._assignment_statement()
        else:
            return self._empty()

    def _assignment_statement(self):
        lvalue = self._variable()
        token = self._current_token
        self._eat(TokenTypes.ASSIGN)
        rvalue = self._expr()
        return AssignNode(lvalue, token, rvalue)

    def _variable(self):
        return VarNode(self._eat(TokenTypes.ID))

    def _empty(self):
        return NoOpNode()

    def _expr(self):
        result = self._term()
        while self._current_token.token_type in [TokenTypes.PLUS,
                                                 TokenTypes.MINUS]:
            op = self._eat()
            result = BinaryOpNode(result, op, self._term())
        return result

    def _term(self):
        """term : factor ((MUL | INTEGER_DIV | FLOAT_DIV) factor)*"""
        result = self._factor()
        while self._current_token.token_type in [TokenTypes.MULTIPLY,
                                                 TokenTypes.INTEGER_DIV,
                                                 TokenTypes.FLOAT_DIV]:
            op = self._eat()
            result = BinaryOpNode(result, op, self._factor())
        return result

    def _factor(self):
        """factor : PLUS factor | MINUS factor
                  | INTEGER_CONST | REAL_CONST
                  | LPAREN expr RPAREN | variable"""
        if self._current_token.token_type in [TokenTypes.PLUS,
                                              TokenTypes.MINUS]:
            op = self._eat()
            return UnaryOpNode(op, self._factor())
        elif self._current_token.token_type == TokenTypes.INTEGER_CONST:
            token = self._eat()
            token.value = int(token.value)
            return NumNode(token)
        elif self._current_token.token_type == TokenTypes.REAL_CONST:
            token = self._eat()
            token.value = float(token.value)
            return NumNode(token)
        elif self._current_token.token_type == TokenTypes.LPAREN:
            self._eat(TokenTypes.LPAREN)
            node = self._expr()
            self._eat(TokenTypes.RPAREN)
            return node
        else:
            return self._variable()

    def parse(self):
        self._current_token = self._lexer.get_next_token()
        node = self._program()
        self._eat(TokenTypes.EOF)
        return node


class NodeVisiter:
    def visit(self, node):
        method_name = 'visit_' + type(node).__name__
        visitor = getattr(self, method_name, self.generic_visit)
        return visitor(node)

    def generic_visit(self, node):
        raise Exception('No visit_{} method'.format(type(node).__name__))


class Interpreter(NodeVisiter):
    def __init__(self):
        self.GLOBAL_SCOPE = {}

    def visit_NumNode(self, node):
        return node.value

    def visit_UnaryOpNode(self, node):
        inner_value = self.visit(node.child)
        if node.token.token_type == TokenTypes.MINUS:
            return -inner_value
        else:
            return inner_value

    def visit_BinaryOpNode(self, node):
        first_value = self.visit(node.left)
        second_value = self.visit(node.right)
        return {TokenTypes.PLUS: lambda x, y: x+y,
                TokenTypes.MINUS: lambda x, y: x-y,
                TokenTypes.MULTIPLY: lambda x, y: x*y,
                TokenTypes.INTEGER_DIV: lambda x, y: x//y,
                TokenTypes.FLOAT_DIV: lambda x, y: x/y,
                }[node.token.token_type](first_value, second_value)

    def visit_ProgramNode(self, node):
        self.visit(node.block)

    def visit_BlockNode(self, node):
        self.visit(node.child)

    def visit_CompoundNode(self, node):
        for child in node.children:
            self.visit(child)

    def visit_AssignNode(self, node):
        var_name = node.lvalue.token.value
        self.GLOBAL_SCOPE[var_name] = self.visit(node.rvalue)

    def visit_VarNode(self, node):
        var_name = node.token.value
        if var_name not in self.GLOBAL_SCOPE:
            raise Exception("{} used before being assigned".format(var_name))
        else:
            return self.GLOBAL_SCOPE[var_name]

    def visit_NoOpNode(self, node):
        pass


def interpret_and_print(text):
    ast = Parser(Lexer(text)).parse()
    interpreter = Interpreter()
    interpreter.visit(ast)
    print(interpreter.GLOBAL_SCOPE)


def main():
    if len(sys.argv) > 1:
        with open(sys.argv[1], 'r') as f:
            interpret_and_print(f.read())
    else:
        while True:
            try:
                text = input("spi> ")
            except EOFError:
                break
            if not text:
                continue
            interpret_and_print(text)

if __name__ == '__main__':
    main()

File: spi.rb
class Token
	attr_accessor :value
	attr_accessor :type
	def initialize(type, value)
		raise "Unknown type" unless [
			:INTEGER_CONST,:PLUS,:MINUS,:MULTIPLY,:FLOAT_DIVIDE,:PROGRAM,:SEMI,
			:DOT,:VAR,:ID,:COLON,:REAL_CONST,:INTEGER_DIVIDE,:LPAREN,:RPAREN,
			:BEGIN,:END,:ASSIGN,:COMMA,:COLON,:INTEGER,:REAL,:EOF].include?(type)
		@type = type
		@value = value
	end

	def inspect
		"<Token #{@type}: #{@value}>"
	end

	def to_s
		inspect
	end
end

class Lexer
	def initialize(text)
		@text = text
	end

	def get_next_token
		remove_whitespace_and_comments

		match = /^(\d+\.\d+)/.match @text
		unless match.nil?
			@text = match.post_match
			return Token.new :REAL_CONST, match.to_s.to_f
		end

		match = /^(\d+)/.match @text
		unless match.nil?
			@text = match.post_match
			return Token.new :INTEGER_CONST, match.to_s.to_i
		end

		match = /^(\+)/.match @text
		unless match.nil?
			@text = match.post_match
			return Token.new :PLUS, match.to_s
		end

		match = /^(-)/.match @text
		unless match.nil?
			@text = match.post_match
			return Token.new :MINUS, match.to_s
		end

		match = /^(\*)/.match @text
		unless match.nil?
			@text = match.post_match
			return Token.new :MULTIPLY, match.to_s
		end

		match = /^(\/)/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:FLOAT_DIVIDE, match.to_s)
		end

		match = /^(DIV)/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:INTEGER_DIVIDE, match.to_s)
		end

		match = /^(;)/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:SEMI, match.to_s)
		end

		match = /^(,)/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:COMMA, match.to_s)
		end

		match = /^(:=)/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:ASSIGN, match.to_s)
		end

		match = /^(:)/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:COLON, match.to_s)
		end

		match = /^(\.)/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:DOT, match.to_s)
		end

		match = /^(\()/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:LPAREN, match.to_s)
		end

		match = /^(\))/.match(@text)
		unless match.nil?
			@text = match.post_match
			return Token.new(:RPAREN, match.to_s)
		end

		match = /^(_?[A-z][A-z0-9]*)/.match(@text)
		unless match.nil?
			@text = match.post_match
			id = match.to_s
			type = :ID
			case id
				when /^VAR$/i then type = :VAR
				when /^PROGRAM$/i then type = :PROGRAM
				when /^BEGIN$/i then type = :BEGIN
				when /^END$/i then type = :END
				when /^INTEGER$/i then type = :INTEGER
				when /^REAL$/i then type = :REAL
			end
			return Token.new(type, id)
		end

		if @text.empty?
			return Token.new(:EOF, nil)
		end

		raise "Unable to parse begining of " + @text
	end

	private

	def remove_whitespace_and_comments
		loop do
			@text = @text.lstrip
			if @text[0] == '{'
				match = /^\{[^}]*\}/.match(@text)
				if match.nil?
					raise "Unmatched comment starting at #{@text}"
				end
				@text = match.post_match
			else
				break
			end
		end
	end
end

class ASTNode
	attr_accessor :token
	def initialize(token)
		@token = token
	end

	def to_s
		inspect
	end

	def inspect
		"<#{self.class.name} token=#{@token.type} value=#{@token.value}>"
	end
end

class NumNode < ASTNode
end

class BinOpNode < ASTNode
	attr_accessor :left, :right
	def initialize(token, left, right)
		super(token)
		@left = left
		@right = right
	end

	def inspect
		"<#{self.class.name} op=#{@token.type} left=#{@left} right=#{@right}>"
	end
end

class VarNode < ASTNode
end

class UnaryOpNode < ASTNode
	attr_accessor :child
	def initialize(token, child)
		super(token)
		@child = child
	end

	def inspect
		"<#{self.class.name} op=#{@token.type} child=#{@child}>"
	end
end

class AssignmentNode < ASTNode
	attr_accessor :lhs, :rhs
	def initialize(lhs, rhs)
		super(nil)
		@lhs = lhs
		@rhs = rhs
	end

	def inspect
		"<#{self.class.name} lhs=#{@lhs} rhs=#{@rhs}>"
	end
end

class TypeNode < ASTNode
	attr_accessor :type
	def initialize(token)
		super(token)
		@type = token.value.upcase
	end
end

class VariableDeclarationNode < ASTNode
	attr_accessor :variable, :type_node
	def initialize(variable_node, type_node)
		super(nil)
		@variable = variable_node
		@type_node = type_node
	end

	def inspect
		"<#{self.class.name} var=#{@variable.value} type=#{@type_node.type}>"
	end
end

class BlockNode < ASTNode
	attr_accessor :declarations, :statements
	def initialize(declarations, statements)
		super(nil)
		@declarations = declarations
		@statements = statements
	end

	def inspect
		"<#{self.class.name} vars=#{@declarations} code=#{@statements}>"
	end
end

class ProgramNode < ASTNode
	attr_accessor :program_name, :block
	def initialize(program_name, block)
		super(nil)
		@program_name = program_name
		@block = block
	end

	def inspect
		"<#{self.class.name} name=#{@program_name.token.value} program=#{@block}>"
	end
end

class Parser
	def initialize(lexer)
		@lexer = lexer
		@current_token = lexer.get_next_token
	end

	def parse
		output = program
		eat :EOF
		output
	end

	private

	def eat(token_type=nil)
		error_msg = "Expected #{token_type} but current token is #{@current_token}"
		if token_type.respond_to?('include?')
			raise error_msg unless token_type.include? @current_token.type
		elsif not token_type.nil?
			raise error_msg unless token_type == @current_token.type
		end

		output = @current_token
		@current_token = @lexer.get_next_token
		output
	end

	def variable
		# variable : ID
		VarNode.new eat :ID
	end

	def factor
		# factor : PLUS factor
		#        | MINUS factor
		#        | INTEGER_CONST
		#        | REAL_CONST
		#        | LPAREN expr RPAREN
		#        | variable
		case @current_token.type
		when :PLUS, :MINUS
			UnaryOpNode.new eat, factor
		when :INTEGER_CONST, :REAL_CONST
			NumNode.new eat
		when :LPAREN
			eat(:LPAREN)
			output = expr
			eat(:RPAREN)
			output
		when :ID
			variable
		end
	end

	def term
		# term : factor ((MULTIPLY | INTEGER_DIVIDE | FLOAT_DIVIDE) factor)*
		output = factor
		while [:MULTIPLY,:INTEGER_DIVIDE,:FLOAT_DIVIDE].include? @current_token.type
			output = BinOpNode.new eat, output, factor
		end
		output
	end

	def expr
		# expr : term ((PLUS | MINUS) term)*
		output = term
		while [:PLUS,:MINUS].include? @current_token.type
			output = BinOpNode.new eat, output, term
		end
		output
	end

	def empty
		# Returns empty array without any parsing
		[]
	end

	def assignment_statement
		# assignment_statement : variable ASSIGN expr
		var = variable
		eat :ASSIGN
		AssignmentNode.new var, expr
	end

	def statement
		# returns array of AssignmentNode objects
		# statement : compound_statement
		#           | assignment_statement
		#           | empty
		case @current_token.type
		when :BEGIN
			return compound_statement
		when :ID
			return [assignment_statement]
		else
			return empty
		end
	end

	def statement_list
		# statement_list : statement
		#                | statement SEMI statement_list
		output = statement
		if @current_token.type == :SEMI
			eat
			output += statement_list
		end
		output
	end

	def compound_statement
		# returns array of AssignmentNode objects
		# compound_statement : BEGIN statement_list END
		eat :BEGIN
		output = statement_list
		eat :END
		output
	end

	def type_spec
		# type_spec : REAL | INTEGER
		TypeNode.new eat([:REAL,:INTEGER])
	end

	def variable_declaration
		# returns Array of VariableDeclarationNode objects
		# variable_declaration : ID (COMMA ID)* COLON type_spec
		ids = [eat(:ID)]
		while @current_token.type == :COMMA
			eat
			ids << eat(:ID)
		end
		eat :COLON
		type = type_spec

		output = []
		for id in ids
			output << VariableDeclarationNode.new(id, type)
		end
		output
	end

	def declarations
		# returns Array of VariableDeclarationNode objects
		# declarations : VAR (variable_declaration SEMI)+
		#              | empty
		if @current_token.type == :VAR
			eat
			output = variable_declaration
			eat :SEMI
			while @current_token.type == :ID
				output += variable_declaration
				eat :SEMI
			end
			return output
		else
			return empty
		end
	end

	def block
		# block : declarations compound_statement
		BlockNode.new declarations, compound_statement
	end

	def program
		# program : PROGRAM variable SEMI block DOT
		eat :PROGRAM
		name = variable
		eat :SEMI
		output = ProgramNode.new name, block
		eat :DOT
		return output
	end
end

class NodeVisiter
	def visit(node)
		node_name = node.class.name
		self.public_send "visit_" + node_name, node
	end
end

class Interpreter < NodeVisiter
	def initialize(ast_tree)
		@ast_tree = ast_tree
		@GLOBAL = Hash.new
	end

	def interpret
		visit @ast_tree
	end

	def visit_ProgramNode(node)
		puts "---------Program: #{node.program_name.token.value}---------"
		visit node.block
	end

	def visit_BlockNode(node)
		puts "declarations:"
		for decl in node.declarations
			visit decl
		end

		puts "Evaluating statements:"
		for statement in node.statements
			visit statement
		end
	end

	def visit_VariableDeclarationNode(node)
		puts "\t#{node.variable.value} : #{node.type_node.type}"
	end

	def visit_AssignmentNode(node)
		variable_name = node.lhs.token.value
		rhs = visit node.rhs
		puts "...Assigning #{variable_name} to #{rhs}"
		@GLOBAL[node.lhs.token.value] = rhs
	end

	def visit_NumNode(node)
		node.token.value
	end

	def visit_VarNode(node)
		variable_name = node.token.value
		unless @GLOBAL.include?(variable_name)
			raise "Variable #{variable_name} not found"
		end
		@GLOBAL[variable_name]
	end

	def visit_BinOpNode(node)
		case node.token.type
		when :PLUS then visit(node.left) + visit(node.right)
		when :MINUS then visit(node.left) - visit(node.right)
		when :MULTIPLY then visit(node.left) * visit(node.right)
		when :FLOAT_DIVIDE then visit(node.left).to_f / visit(node.right)
		when :INTEGER_DIVIDE then (visit(node.left) / visit(node.right)).to_i
		else raise "Unknown Binary Operations: #{node}"
		end
	end

	def visit_UnaryOpNode(node)
		case node.token.type
		when :PLUS then visit(node.child)
		when :MINUS then -visit(node.child)
		else raise "Unknown Unary Operations: #{node}"
		end
	end
end

filename = ARGV.first
txt = open(filename).read
Interpreter.new(Parser.new(Lexer.new(txt)).parse()).interpret
